src/librustc_trans/trans/base.rs

```diff
@@ -775,15 +803,33 @@ pub fn load_if_immediate<'blk, 'tcx>(cx: Block<'blk, 'tcx>,
 /// gives us better information about what we are loading.
 pub fn load_ty<'blk, 'tcx>(cx: Block<'blk, 'tcx>,
                            ptr: ValueRef, t: Ty<'tcx>) -> ValueRef {
+    use trans::basic_block::BasicBlock;
+
     if cx.unreachable.get() || type_is_zero_size(cx.ccx(), t) {
         return C_undef(type_of::type_of(cx.ccx(), t));
     }
 
     let ptr = to_arg_ty_ptr(cx, ptr, t);
+    let ptr = Value(ptr);
     let align = type_of::align_of(cx.ccx(), t);
 
+    let bb = BasicBlock(cx.llbb);
+
     if type_is_immediate(cx.ccx(), t) && type_of::type_of(cx.ccx(), t).is_aggregate() {
-        let load = Load(cx, ptr);
+        if let Some(val) = ptr.get_dominating_store(cx) {
```

Comment:
Oops, sent too early. Using a function would avoid the code duplication with the same block below, and would allow us to keep using a `ValueRef` in this function, saving all the `.get()` noise.